### PR TITLE
VEN-405 | File Upload for Harbor

### PR DESCRIPTION
**Pull request by a-rmz**
## Description :sparkles:
- Replace the `image_file` input field for `Harbor` mutations (`create` and `update`) with File `Upload` scalar

## Issues :bug:
**Closes 🙅‍♂** 
**[VEN-405](https://helsinkisolutionoffice.atlassian.net/browse/VEN-405)**

**Related 🤝**
[**[VEN-512](https://helsinkisolutionoffice.atlassian.net/browse/VEN-512)**] Add tests for file upload

## Testing :alembic:
**Automated tests ⚙️**
N/A (Refer to VEN-512)

**Manual testing 👷**
First, you have to comment/remove the `@login_required` and `@superuser_required` decorators from the [update mutation](https://github.com/City-of-Helsinki/berth-reservations/blob/ven-405/file-upload/resources/schema.py#L510) to avoid having to deal with login with cURL.

To perform an update, execute the following cURL:
```shell
 curl localhost:8000/graphql_v2/ \
  -H "Content-Type: multipart/form-data" \
  -F operations='{ "query": "mutation UpdateHarbor($updateHarbor: UpdateHarborMutationInput!) { updateHarbor(input: $updateHarbor) { harbor { id properties { imageFile name } } } }", "variables": { "updateHarbor": { "id": "SGFyYm9yTm9kZTo3OTIyZTBiZC0wNDNhLTRmMjEtYjU0ZC05ZjU5NGU1ZGNhZDY=", "imageFile": null  } } }' \
  -F map='{ "0": ["variables.updateHarbor.imageFile"] }' \
  -F 0=@relative-path-to-image.png
```

It will perform the following mutation, replacing the `null`-ed variable (`imageFile`) with the file passed. For a detailed explanation, refer to the **Additional notes** section.
```graphql
mutation UpdateHarbor($updateHarbor: UpdateHarborMutationInput!) {
  updateHarbor(input: $updateHarbor) {
    harbor {
      id
      properties {
        imageFile
        name
      }
    }
  }
}
```

## Screenshots :camera_flash:
**Description in Harbor mutations**
image

## Additional notes :spiral_notepad:
The file upload follows the [GraphQL Multipart Request Spec](https://github.com/jaydenseric/graphql-multipart-request-spec), so any of the [clients that implement it](https://github.com/jaydenseric/graphql-multipart-request-spec#client) can be used to interact with it.

That document explains why [requests have that shape](https://github.com/jaydenseric/graphql-multipart-request-spec#multipart-form-field-structure) which requires `operations`, `map`, and `files`.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -7,32 +7,33 @@
...
-django-cors-headers==3.1.1
-django-enumfields==1.0.0
-django-environ==0.4.5
-django-filter==2.2.0
-django-graphql-geojson==0.1.4
-django-graphql-jwt==0.2.2
-django-helusers==0.5.2
-django-ilmoitin==0.2.0
+django-anymail==7.0.0     # via -r requirements.in, django-ilmoitin
+django-cors-headers==3.1.1  # via -r requirements.in
```
Comment (igordavydsson) [inline requirements.txt]: how did you get all of these `# via -r requirements.in`? 🤔

File: requirements.in [hunk 2]
```diff
@@ -14,6 +14,7 @@ django-parler
 django-ilmoitin
 factory-boy
 graphene-django
+graphene-file-upload
```
Comment (igordavydsson) [inline requirements.in]: I think, alphabetically it should come after `graphene-federation` 😅

File: requirements.txt (hunk 1 above)
Comment (a-rmz) [reply requirements.txt]: Just by running `pip-compile requirements.in`. I had some trouble with the whole `venv` seutp, so I had to run it with the absolute path from the venv bin directory, so idk if that could've affected somehow...

If you remove the `requirements.txt` and run `pip-compile`, do you get it with or without those comments? Maybe it's a newer version?

File: requirements.in (hunk 2 above)
Comment (a-rmz) [reply requirements.in]: 😭 so close yet so far, I thought I had gotten it right.

File: requirements.txt (hunk 1 above)
Comment (igordavydsson) [reply requirements.txt]: hm, interesting. I also get now these `via -r requirements.in`. I guess it's the way it should be with the primary dependencies now. Change was apparently introduced in [pip-tools v4.5.0](https://github.com/jazzband/pip-tools/blob/4.5.1/CHANGELOG.md#450-2020-02-20)

Comment (a-rmz) [reply requirements.txt]: Good catch, I kinda like it better, it gives you a bit more `T - Transparency` on what comes from where.

**Pull request merged**